src/libcore/xml.cpp

```diff
@@ -366,6 +369,20 @@ void upgrade_tree(XMLSource &src, pugi::xml_node &node, const Version &version)
         }
         for (pugi::xpath_node result: node.select_nodes("//lookAt"))
             result.node().set_name("lookat");
+        // automatically rename reserved identifiers
+        for (pugi::xpath_node result: node.select_nodes("//@id")) {
+            pugi::xml_attribute id_attrib = result.attribute();
+            char const* val = id_attrib.value();
+            if (val && val[0] == '_') {
+                std::string new_id = std::string("ID") + val + "__UPGR";
+                Log(Warn, "Changing identifier: \"%s\" -> \"%s\"", val, new_id.c_str());
+                id_attrib = new_id.c_str();
+            }
+        }
```

Comment:
Could you illustrate this by an example maybe? I am not sure to see

Reply:
Excerpt from the pool scene: https://cgg.mff.cuni.cz/~jirka/papers/2014/olpm/index.htm

```
	<bsdf type="phong" id="_01-Default">
		<rgb name="diffuseReflectance" value="0.67451 0.67451 0.67451"/>
		<rgb name="specularReflectance" value="0.0128 0.0128 0.0128"/>
		<float name="exponent" value="4"/>
	</bsdf>
```

With the patch, it is allowed to load, giving:
```
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_02-Default" -> "ID_02-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_01-Default" -> "ID_01-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_02-Default" -> "ID_02-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_01-Default" -> "ID_01-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_01-Default" -> "ID_01-Default__UPGR"
```

Without the patch, it crashes:
```
            // src/libcore/xml.cpp
            if (string::starts_with(id, "_"))
                src.throw_error(
                    node, "invalid id \"%s\" in element \"%s\": leading "
                          "underscores are reserved for internal identifiers.",
                    id, node.name());
```

Reply:
Thanks for your explanation, makes more sense now.